### PR TITLE
[FSDP][7/N] Add alignment padding for `use_orig_params=True`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97667
* #97666
* #97665
* #97664
* #97663
* #97662
* #97661

This PR adds intra-`FlatParameter` 16-byte alignment padding to the `use_orig_params=True` code path to avoid clones in TorchInductor.

**Approach**
The `FlatParameter` maintains several data structures about its original parameters. Notably, the data structures `_param_infos`, `_shapes`, `_numels`, and `_fqns` have the same length and index in the same way.

This PR treats alignment padding _like_ an original parameter in that the padding gets flattened into the `FlatParameter`. Therefore, it must be reflected in the aforementioned data structures. However, given the way in which the data structures are used, we choose to do the following if the `i`th tensor flattened into the `FlatParameter` is padding:
- `_numels[i]` is the numel of padding
- `_param_infos[i] == _shapes[i] == _fqns[i] == None`

This choice is because (1) we must record the padding numel to account for it (e.g. for views) and (2) we prefer to preserve the invariant that the data structures index in the same way over avoiding `None` entries.

To ease the burden of other FSDP developers, we separate the parameter flattening logic:
- `_init_flat_param_and_metadata()`: This should be called only once in the `FlatParamHandle` constructor. The `FlatParameter` metadata is assumed to be static thereafter.
- `flatten_tensors()` / `flatten_tensors_into_flat_param()`: These can be used for optimizer and model state dict and can be called after construction time.

This separation allows `_init_flat_param_and_metadata()` to contain the much heavier metadata logic, while keeping the latter methods to be much lighter. The only constraint is that the alignment padding logic must be kept consistent between the two, but this should be worth the simper interface.

**Testing**
- This PR directly modifies the `use_orig_params=True` code path, so all existing tests passing gives good signal.
    - Some existing unit tests had to be adjusted to account for the alignment padding.
- This PR adds two tests in `test_fsdp_flatten_params.py` to explicitly test the sharding metadata with alignment for both parameter full precision and mixed precision since the latter requires possibly more padding elements due to the decreased per-element size.

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_ignored_modules.py [hunk 1]
```diff
@@ -146,7 +146,15 @@ def _test_ignored_modules_transformer(
         with FSDP.summon_full_params(wrapped_model):
             flat_param = wrapped_model.params[0]
             flat_param_numel = flat_param.numel()
-            self.assertEqual(flat_param_numel, nonignored_numel)
+            if use_orig_params:
+                # Subtract numel from alignment padding
+                padding_numel = sum(
+                    numel
+                    for (numel, pi) in zip(flat_param._numels, flat_param._param_infos)
+                    if pi is None
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_ignored_modules.py:154]: could we ever have `None` in _param_infos if it's not a padding phantom parameter? This might be a bit risky, could we have some global constant to indicate this is a padding parameter and not a real parameter?

File: test/distributed/fsdp/test_fsdp_ignored_modules.py [hunk 2]
```diff
@@ -185,6 +193,15 @@ def _test_ignored_modules_nested(self, use_orig_params: bool, ignore_modules: bo
         with FSDP.summon_full_params(wrapped_model):
             flat_param = wrapped_model.params[0]
             flat_param_numel = flat_param.numel()
+            if use_orig_params:
+                # Subtract numel from alignment padding
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_ignored_modules.py:197]: similar comment as above

File: test/distributed/fsdp/test_fsdp_use_orig_params.py [hunk 3]
```diff
@@ -411,9 +411,9 @@ def _test_multiple_optimizers(self, sharding_strategy: ShardingStrategy):
                 has_weight = False
                 has_bias = False
                 for param, fqn in zip(flat_param._params, flat_param._fqns):
-                    if "weight" in fqn and param.numel() > 0:
+                    if fqn is not None and "weight" in fqn and param.numel() > 0:
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_use_orig_params.py:414]: Is there any way _params and _fqns could just not return these padding parameters which aren't real parameters? i.e. have a `@property` accessor that filters these out?

File: torch/distributed/fsdp/_common_utils.py [hunk 4]
```diff
@@ -212,7 +212,9 @@ def module_fn(module, prefix, param_to_fqns):
                 else [param_name]
             )  # prefixed from `module`
             global_fqns = [
-                clean_tensor_name(prefix + name) for name in local_fqns
+                clean_tensor_name(prefix + name)
+                for name in local_fqns
+                if name is not None
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_common_utils.py:217]: same, can we just make local_fqns not have None when we read it?

File: torch/distributed/fsdp/_optim_utils.py [hunk 5]
```diff
@@ -264,11 +265,22 @@ def _unflatten_communicated_optim_state(
         for state_name, flat_tensor in sorted_items(tensor_state):
             views_generated = state_name in flat_param_views
             if not views_generated:
-                views = FlatParamHandle._get_unflat_views(flat_param, flat_tensor)
+                views = handle._get_unflat_views(flat_tensor)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_optim_utils.py:268]: is this just a change from  using it as a classmethod to instance method?

File: torch/distributed/fsdp/_optim_utils.py [hunk 6]
```diff
@@ -264,11 +265,22 @@ def _unflatten_communicated_optim_state(
...
                 flat_param_views[state_name] = views
             else:
                 views = flat_param_views[state_name]
-            optim_state: Union[torch.Tensor, ShardedTensor] = next(views)
+            optim_state: Optional[Union[torch.Tensor, ShardedTensor]] = next(views)
+            # Skip any alignment padding -- `views` should never be exhausted
+            # before the outer for loop completes
+            try:
+                while optim_state is None:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_optim_utils.py:276]: are we sure we can never get `None` other than alignment padding?

File: torch/distributed/fsdp/flat_param.py [hunk 7]
```diff
@@ -2100,6 +2286,46 @@ def _safe_setattr_tensor_or_param(
...
+def _construct_padding_tensor(
+    padding_numel: int, dtype: torch.dtype, requires_grad: bool, device: torch.device
+):
+    # NOTE: Set the padding value as a magic number for debuggability. The
+    # value itself should never be used in any user-facing computation.
+    return (
+        torch.ones(
+            (padding_numel,), dtype=dtype, requires_grad=requires_grad, device=device
+        )
+        * 42
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:2325]: nit: use a constant?

File: torch/distributed/fsdp/flat_param.py [hunk 8]
```diff
@@ -2100,6 +2286,46 @@ def _safe_setattr_tensor_or_param(
     setattr(module, param_name, tensor_or_param)
 
 
+def _convert_to_params(
+    tensors: List[Union[torch.Tensor, nn.Parameter]]
+) -> List[nn.Parameter]:
+    return [t if isinstance(t, nn.Parameter) else nn.Parameter(t) for t in tensors]
+
+
+def _detach_if_needed(param_or_tensor: Union[nn.Parameter, Tensor]) -> Tensor:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:2295]: nit: would this be better in a utils file?

File: torch/distributed/fsdp/flat_param.py [hunk 9]
```diff
@@ -2100,6 +2286,46 @@ def _safe_setattr_tensor_or_param(
     setattr(module, param_name, tensor_or_param)
 
 
+def _convert_to_params(
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:2289]: would this be better in a utils file?

File: torch/distributed/fsdp/flat_param.py [hunk 10]
```diff
@@ -354,67 +366,98 @@ def __init__(
...
+            )
+        self._init_setattr_fns()
+        align_addresses = use_orig_params
+        self._init_get_unflat_views_fn(align_addresses)
+
         self.device = device
         self.process_group = process_group
         self.rank = process_group.rank()
         self.world_size = process_group.size()
+
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:382]: nit: remove?

File: torch/distributed/fsdp/flat_param.py [hunk 11]
```diff
@@ -354,67 +366,98 @@ def __init__(
...
-        self._orig_param_dtype = self.flat_param.dtype
-        self._use_unsharded_views(as_params=False)
+
+        # Optimistically assume a valid input `params` and set dtype attributes
+        # before `_init_flat_param()`, which performs the actual validation
+        self._orig_param_dtype = params[0].dtype
         self._init_param_reduce_dtypes(mp_param_dtype, mp_reduce_dtype)
+        assert self._fwd_bwd_param_dtype is not None  # mypy
+        self._aligned_numel = (
+            _get_aligned_numel(unsharded_dtype=self._fwd_bwd_param_dtype)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:397]: why do we have to distinguish btwn unsharded_dtype vs sharded_dtype here, wouldn't they be the same?

File: torch/distributed/fsdp/flat_param.py [hunk 12]
```diff
@@ -354,67 +366,98 @@ def __init__(
...
         """
-        Initializes the flat parameter ``self.flat_param`` by flattening the
-        parameters in ``params`` into a single :class:`FlatParameter` and saves
-        relevant metadata. Shared parameters are only included in the flat
-        parameter once.
-
-        This checks that all comprising parameters have the same dtype and
-        ``requires_grad`` and does not support nested construction of
-        :class:`FlatParameter` s.
+        NOTE: This should only be called once at construction time, after which
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:432]: unrelated to this PR, but can we enforce this in code?

File: torch/distributed/fsdp/flat_param.py [hunk 13]
```diff
@@ -466,65 +504,131 @@ def _init_flat_param(
...
+            if type(tensor) is FlatParameter:
+                raise ValueError("Cannot flatten a `FlatParameter`")
+            if dtype is None and not tensor.is_floating_point():
+                raise ValueError("Cannot flatten integer dtype tensors")
+            if dtype is not None and tensor.dtype != dtype:
+                raise ValueError(
+                    f"Must flatten tensors with uniform dtype but got {dtype} "
+                    f"and {tensor.dtype}"
+                )
+            # TODO: Relax the following for `use_orig_params=True`.
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:560]: does this mean use_orig_params=True does not technically support different requires_grad within a flat param currently? can we file an issue?

File: torch/distributed/fsdp/flat_param.py [hunk 14]
```diff
@@ -466,65 +504,131 @@ def _init_flat_param(
...
+            if requires_grad is not None and tensor.requires_grad != requires_grad:
+                raise ValueError("Must flatten tensors with uniform `requires_grad`")
+            if device is not None and tensor.device != device:
+                raise ValueError(
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:564]: unrelated to this PR, but don't we check same device at the module level as well in FSDP, do we still need to do so here?

File: torch/distributed/fsdp/flat_param.py [hunk 15]
```diff
@@ -1356,41 +1484,63 @@ def _use_sharded_flat_param(self) -> None:
...
 
+    @no_type_check
+    def _get_unflat_views_aligned(
+        self,
+        tensor: Optional[Tensor] = None,
+    ) -> List[Optional[Tensor]]:
+        """
+        This has the same contract as :meth:`_get_unflat_views_unaligned`
+        except it checks for ``None`` placeholders representing padding for
+        alignment, which may incur slightly more CPU overhead.
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:1521]: nit: add docs about the input args?

**Review by rohan-varma — APPROVED** [review]
LGTM overall, thanks for pushing this through quickly!

Main comment is around whether we can consolidate the code that checks if a parameter is actually just padding in several places, and maybe use a global constant instead of `None`, and ensure callsites that just want regular parameter info won't have to know about / do this None filtering?

File: torch/distributed/fsdp/flat_param.py (hunk 14 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:564]: Good point! I think the check here is okay since then this method is self-contained.

The way I think about it:
- The check here is necessary for us to correctly flatten the parameters without erroring (or else we are passing different devices into `torch.cat()`).
- The check in the FSDP constructor is for us to determine what is the `compute_device`, which is not well-defined if we have parameters on multiple devices.

File: torch/distributed/fsdp/flat_param.py (hunk 15 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:1521]: I am wondering, should we repeat the comment from `_get_unflat_views_unaligned()` or just refer developers to it?

I worry about maintaining two sets of comments.

File: torch/distributed/fsdp/flat_param.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:2289]: Yes, we can move it to a utils file.

File: torch/distributed/fsdp/flat_param.py (hunk 8 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:2295]: Yes, we can move it to a utils file.

File: torch/distributed/fsdp/flat_param.py [hunk 16]
```diff
@@ -2100,6 +2286,46 @@ def _safe_setattr_tensor_or_param(
...
+
+def _detach_if_needed(param_or_tensor: Union[nn.Parameter, Tensor]) -> Tensor:
+    return (
+        param_or_tensor.detach()
+        if isinstance(param_or_tensor, nn.Parameter)
+        else param_or_tensor
+    )
+
+
+def _get_aligned_numel(unsharded_dtype: torch.dtype):
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:2303]: Sorry if there was any confusion. The return type is in units of numel, not bytes (since both `ALIGNMENT` and `unsharded_dtype_size` are in units of bytes, which means that they cancel).

File: torch/distributed/fsdp/flat_param.py (hunk 7 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:2325]: Sounds good!

File: test/distributed/fsdp/test_fsdp_ignored_modules.py [hunk 17]
```diff
@@ -146,7 +146,15 @@ def _test_ignored_modules_transformer(
         with FSDP.summon_full_params(wrapped_model):
             flat_param = wrapped_model.params[0]
             flat_param_numel = flat_param.numel()
-            self.assertEqual(flat_param_numel, nonignored_numel)
+            if use_orig_params:
+                # Subtract numel from alignment padding
```
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_ignored_modules.py:150]: I realized I meant "subtract the numel that comes from the alignment padding" -- let me clarify this

File: torch/distributed/fsdp/flat_param.py (hunk 13 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:560]: https://github.com/pytorch/pytorch/issues/91167

File: test/distributed/fsdp/test_fsdp_ignored_modules.py (hunk 1 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_ignored_modules.py]: I refactored the PR so that we have two versions of each data structure that can contain a padding element. This means that the developer will need to make an explicit decision which version to use. Hopefully this helps.

File: torch/distributed/fsdp/flat_param.py [hunk 18]
```diff
@@ -152,46 +168,77 @@ class FlatParameter(nn.Parameter):
         e.g. to the sharded flat parameter, low precision sharded flat
         parameter, or the unsharded flat parameter.
 
+    NOTE: Since ``use_orig_params=True`` supports intra-``FlatParameter``
+    padding, we save tuples/lists that include padding objects and tuples/lists
+    with them filtered. The former are prefixed with "_wp", which stands for
+    "with padding", and should mainly be used within this file. They are needed
+    when iterating over the tensors in the ``FlatParameter`` or analogous
+    tensors and the index matters. Developers should take care to *avoid*
```
Comment (awgu) [inline torch/distributed/fsdp/flat_param.py:176]:
```suggestion
    tensors and the enumerated index matters. Developers should take care to *avoid*
```
